### DISCUSSION TITLE
Query vs. DML vs DDL

### DISCUSSION THREAD

**Discussion by manticore-projects** [JSQLParser/JSqlParser#1290] [Ideas]
Greetings.

At the moment all Statements just extend `Statement` and we do not know if the Statement is a Query, a DML or a DDL (unless we test a lot of `instance of`).

I would like to suggest to
a) either introduce new Abstract Classes `QueryStatement`, `DmlStatement' and `DdlStatement` use these (instead of `Statement`) for implementing each Statement
b) or to declare `Statement` abstract with additional methods `getStatementType()` (returning `Query`, `DDL` or `DML`) and `isQuery()`, `isDml()` and `isDdl()`

Which way would be preferred? I tended to option a).

**Comment by wumpz** [comment]
Wouldn't it be sufficient to introduce some kind of grouping interfaces? What should be within the abstract base clases?

**Comment by manticore-projects** [comment]
Quick and dirty draft for discussion:
```java
public abstract class StatementImpl implements Statement {
    abstract StringBuilder appendTo(StringBuilder builder);
    
    @Override
    public String toString() {
        return appendTo(new StringBuilder()).toString();
    }
}

public abstract class DDLStatement extends StatementImpl {

}

public abstract class DMLStatement extends StatementImpl {

}

public abstract class QueryStatement extends StatementImpl {

}
```

And then every Statement class should extend one of these, like:
``` java
public class Select extends QueryStatement {
....
}

public class Insert extends DMLStatement {
....
}
```

**Comment by manticore-projects** [comment]
Solved via PR #1301
